### PR TITLE
fixed bug where all documents are added as high priority on solution …

### PR DESCRIPTION
**Pull request by heejaechang**
…load.

### Customer scenario

Right after solution is opened, VS uses more memory than it needs to.

### Bugs this fixes

N/A

### Workarounds, if any

There is no workaround

### Risk

No risk

### Performance impact

files that belong to same project should be processed together when solution crawler runs.

### Is this a regression from a previous update?

Yes.

### Root cause analysis

the change below (tweak to make solution crawler to respond to text change quicker) broke solution open case. when solution just opened, all document will be marked as text changed (document is added as empty and then changed to have content). with the change below, basically every files became high priority and made high priority useless for solution open.

https://github.com/dotnet/roslyn/commit/f8186c1b34e99288788e32ae8151c4b51fbc8545#diff-1f3bf79660d31a9ad86af6a2ccdd47e0

this moves back to old behavior. text change is no longer high priority. and we get similar effect by always trying to handle active, open files first without checking high priority map.

### How was the bug found?

Dogfooding

### CODE REVIEW

**Comment by heejaechang** [comment]
@shyamnamboodiripad found the bug why files are not grouped to same project.

**Comment by heejaechang** [comment]
@jinujoseph @dotnet/roslyn-ide @mavasani can you take a look?

File: src/EditorFeatures/Test/SolutionCrawler/WorkCoordinatorTests.cs [hunk 1]
```diff
@@ -1151,6 +1191,29 @@ public Task DocumentResetAsync(Document document, CancellationToken cancellation
...
+            public readonly List<DocumentId> DocumentIds = new List<DocumentId>();
+
+            public Task AnalyzeDocumentAsync(Document document, SyntaxNode bodyOpt, InvocationReasons reasons, CancellationToken cancellationToken)
+            {
+                this.DocumentIds.Add(document.Id);
+                return SpecializedTasks.EmptyTask;
+            }
+
+            #region unused 
+            public bool NeedsReanalysisOnOptionChanged(object sender, OptionChangedEventArgs e) => false;
```
Comment (sharwell) [inline src/EditorFeatures/Test/SolutionCrawler/WorkCoordinatorTests.cs]: :memo: All of these should throw `ExceptionUtilities.Unreachable`

**Comment by CyrusNajmabadi** [comment]
Just to verify, but files that are changed because of user editing will still get priority over everything else (including other open documents), right?

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.NormalPriorityProcessor.cs [hunk 2]
```diff
@@ -232,18 +232,15 @@ private IEnumerable<DocumentId> GetPrioritizedPendingDocuments()
...
                             // First the active document
                             var activeDocumentId = this.Processor._documentTracker.GetActiveDocument();
-                            if (activeDocumentId != null && _higherPriorityDocumentsNotProcessed.ContainsKey(activeDocumentId))
+                            if (activeDocumentId != null)
                             {
                                 yield return activeDocumentId;
                             }
 
                             // Now any visible documents
                             foreach (var visibleDocumentId in this.Processor._documentTracker.GetVisibleDocuments())
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.NormalPriorityProcessor.cs]: how is visibility determined?

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.NormalPriorityProcessor.cs [hunk 3]
```diff
@@ -232,18 +232,15 @@ private IEnumerable<DocumentId> GetPrioritizedPendingDocuments()
                         {
                             // First the active document
                             var activeDocumentId = this.Processor._documentTracker.GetActiveDocument();
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.NormalPriorityProcessor.cs]: how is 'active document' determined?

File: src/EditorFeatures/Test/SolutionCrawler/WorkCoordinatorTests.cs (hunk 1 above)
Comment (heejaechang) [reply src/EditorFeatures/Test/SolutionCrawler/WorkCoordinatorTests.cs]: it is reachable. will be called.

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.NormalPriorityProcessor.cs (hunk 2 above)
Comment (heejaechang) [reply src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.NormalPriorityProcessor.cs]: we have visual studio document tracking service that tracks active file.

Comment (heejaechang) [reply src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.NormalPriorityProcessor.cs]: and workspace tracks what files are opened in host.

**Comment by heejaechang** [comment]
text changes on active/open files will get higher priority over other document.

**Comment by heejaechang** [comment]
@Pilchie can I get approval for 15.7?

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.NormalPriorityProcessor.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.NormalPriorityProcessor.cs]: so "Visibility = Open"?  Should this be called GetOpenDocuments then?

**Comment by CyrusNajmabadi** [comment]
> text changes on active/open files will get higher priority over other document.

My question was more:

will the text change in a document get higher priority than *other* open documents.  i.e. if i have 100 docs open, and i edit one, will it still receive the highest priority?  And will it be above the priority of the other *open* files? 

If so, great.  But i just want to confirm.

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.NormalPriorityProcessor.cs (hunk 2 above)
Comment (heejaechang) [reply src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.NormalPriorityProcessor.cs]: so. the tracker service also track visible (editor that is shown to users) as well.

so, priority is 

active file (where focus is on)
visible files (document that is visible to users)
highpirority files (opened files + ones that specifically asked by users)

**Comment by CyrusNajmabadi** [comment]
Great.  Awesome.  Thanks for the clarity.  LGTM.

**Comment by Pilchie** [comment]
Approved - thanks.

**Comment by shyamnamboodiripad** [comment]
Thanks much for this fix HeeJae!!

**Pull request merged**